File: scripts/frame_test.py
#!/usr/bin/env python3
"""
Binary Frame Protocol - Python Reference Implementation + Test Vectors
Phase 1: Verify encoding/decoding correctness
"""

from enum import IntEnum
from typing import List, Dict, Any, Optional, Callable
import struct
import json

class WireType(IntEnum):
    VARINT = 0
    FIXED64 = 1
    LENGTH_DELIMITED = 2
    START_GROUP = 3
    END_GROUP = 4
    FIXED32 = 5

def make_tag(field_num: int, wire_type: WireType) -> int:
    return (field_num << 3) | wire_type

def encode_varint(value: int) -> bytes:
    """Encode unsigned integer as varint."""
    result = bytearray()
    while value > 0x7F:
        result.append((value & 0x7F) | 0x80)
        value >>= 7
    result.append(value & 0x7F)
    return bytes(result)

def decode_varint(data: bytes, offset: int = 0) -> tuple:
    """Decode varint from data at offset. Returns (value, next_offset)."""
    result = 0
    shift = 0
    pos = offset
    while True:
        if pos >= len(data):
            raise ValueError("Incomplete varint")
        byte = data[pos]
        result |= (byte & 0x7F) << shift
        pos += 1
        if not (byte & 0x80):
            break
        shift += 7
        if shift >= 64:
            raise ValueError("Varint too long")
    return result, pos

def encode_field(field_num: int, wire_type: WireType, value) -> bytes:
    """Encode a single field."""
    tag = make_tag(field_num, wire_type)
    result = encode_varint(tag)
    
    if wire_type == WireType.VARINT:
        if isinstance(value, bool):
            result += encode_varint(1 if value else 0)
        else:
            result += encode_varint(value)
    elif wire_type == WireType.LENGTH_DELIMITED:
        if isinstance(value, str):
            value = value.encode('utf-8')
        result += encode_varint(len(value))
        result += value
    else:
        raise ValueError(f"Unsupported wire type: {wire_type}")
    
    return result

def parse_frame(data: bytes) -> tuple:
    """Parse a frame, returns (msg_type, fields_dict)."""
    if len(data) < 1:
        raise ValueError("Empty frame")
    
    msg_type = data[0]
    offset = 1
    fields = {}
    
    while offset < len(data):
        tag, offset = decode_varint(data, offset)
        field_num = tag >> 3
        wire_type = tag & 0x07
        
        if wire_type == WireType.VARINT:
            value, offset = decode_varint(data, offset)
            fields[field_num] = value
        elif wire_type == WireType.LENGTH_DELIMITED:
            length, offset = decode_varint(data, offset)
            if offset + length > len(data):
                raise ValueError("Length-delimited field exceeds frame")
            value = data[offset:offset + length]
            offset += length
            fields[field_num] = value
        else:
            raise ValueError(f"Unsupported wire type in frame: {wire_type}")
    
    return msg_type, fields

class FrameEncoder:
    """Helper to build frames."""
    def __init__(self, msg_type: int):
        self.msg_type = msg_type
        self.data = bytearray()
        self.data.append(msg_type)
    
    def add_varint(self, field_num: int, value: int):
        self.data.extend(encode_field(field_num, WireType.VARINT, value))
        return self
    
    def add_string(self, field_num: int, value: str):
        self.data.extend(encode_field(field_num, WireType.LENGTH_DELIMITED, value))
        return self
    
    def add_bytes(self, field_num: int, value: bytes):
        self.data.extend(encode_field(field_num, WireType.LENGTH_DELIMITED, value))
        return self
    
    def build(self) -> bytes:
        return bytes(self.data)

# ===== 消息类型常量 =====
MSG_HELLO       = 0x01
MSG_STATUS_RPT  = 0x02
MSG_DATA_RPT    = 0x03
MSG_CONFIG_MFST = 0x04
MSG_CONFIG_RSLT = 0x05
MSG_WRITE_CMD   = 0x06
MSG_WRITE_RSP   = 0x07
MSG_PING        = 0x08
MSG_PONG        = 0x09
MSG_OTA_CMD     = 0x0A
MSG_OTA_PROG    = 0x0B
MSG_SCAN_RPT    = 0x0C
MSG_SCAN_REQ    = 0x0D
MSG_QUERY_REQ   = 0x0E
MSG_QUERY_RSP   = 0x0F
MSG_CONFIG_MFST_PY = 0x04  # alias to avoid name clash
MSG_CONFIG_QUERY = 0x10
MSG_CONFIG_REPORT = 0x11
MSG_HELLO_ACK   = 0x12

# ===== 测试向量 =====
def test_hello():
    print("=== Test: Hello ===")
    encoder = FrameEncoder(MSG_HELLO)
    encoder.add_string(1, "AABBCCDDEEFF")
    encoder.add_string(2, "2.0.0")
    encoder.add_string(3, "ESP32S3")
    encoder.add_varint(4, 4)
    frame = encoder.build()
    
    print(f"Encoded ({len(frame)} bytes): {frame.hex()}")
    
    msg_type, fields = parse_frame(frame)
    print(f"Type: 0x{msg_type:02X}")
    print(f"Field 1 (device_id): {fields[1].decode('utf-8')}")
    print(f"Field 2 (firmware): {fields[2].decode('utf-8')}")
    print(f"Field 3 (model): {fields[3].decode('utf-8')}")
    print(f"Field 4 (channel_count): {fields[4]}")
    
    assert msg_type == MSG_HELLO
    assert fields[1] == b"AABBCCDDEEFF"
    assert fields[2] == b"2.0.0"
    assert fields[3] == b"ESP32S3"
    assert fields[4] == 4
    print("PASS")
    return frame

def test_data_report():
    print("\n=== Test: DataReport ===")
    raw = bytes([0x01, 0x02, 0x03, 0x04, 0x05])
    encoder = FrameEncoder(MSG_DATA_RPT)
    encoder.add_varint(1, 1)
    encoder.add_varint(2, 12345678)
    encoder.add_varint(3, 42)
    encoder.add_bytes(4, raw)
    frame = encoder.build()
    
    print(f"Encoded ({len(frame)} bytes): {frame.hex()}")
    
    msg_type, fields = parse_frame(frame)
    print(f"Type: 0x{msg_type:02X}")
    print(f"Field 1 (channel_id): {fields[1]}")
    print(f"Field 2 (timestamp_us): {fields[2]}")
    print(f"Field 3 (sequence): {fields[3]}")
    print(f"Field 4 (raw_data): {fields[4].hex()}")
    
    assert msg_type == MSG_DATA_RPT
    assert fields[1] == 1
    assert fields[2] == 12345678
    assert fields[3] == 42
    assert fields[4] == raw
    print("PASS")
    return frame

def test_varint_edge_cases():
    print("\n=== Test: Varint Edge Cases ===")
    test_cases = [
        (0, "00"),
        (1, "01"),
        (127, "7f"),
        (128, "80 01"),
        (255, "ff 01"),
    ]
    
    for value, expected_hex in test_cases:
        encoded = encode_varint(value)
        expected = bytes.fromhex(expected_hex.replace(" ", ""))
        assert encoded == expected, f"Failed for {value}: got {encoded.hex()}, expected {expected.hex()}"
        decoded, _ = decode_varint(encoded)
        assert decoded == value, f"Decode failed for {value}: got {decoded}"
        print(f"  varint({value}) = {encoded.hex()} OK")
    
    print("PASS")

def test_hello_ack():
    print("\n=== Test: HelloAck ===")
    server_time = 1717200000000  # Unix ms timestamp
    features = 0
    encoder = FrameEncoder(MSG_HELLO_ACK)
    encoder.add_varint(1, server_time)
    encoder.add_varint(2, features)
    frame = encoder.build()
    
    print(f"Encoded ({len(frame)} bytes): {frame.hex()}")
    
    msg_type, fields = parse_frame(frame)
    print(f"Type: 0x{msg_type:02X}")
    print(f"Field 1 (server_time): {fields[1]}")
    print(f"Field 2 (features): {fields[2]}")
    
    assert msg_type == MSG_HELLO_ACK
    assert fields[1] == server_time
    assert fields[2] == features
    print("PASS")
    return frame

if __name__ == "__main__":
    print("=" * 50)
    print("Binary Frame Protocol - Python Reference Implementation")
    print("=" * 50)
    
    test_varint_edge_cases()
    test_hello()
    test_data_report()
    test_hello_ack()
    
    print("\n" + "=" * 50)
    print("All tests PASSED")
    print("=" * 50)

File: backend/pkg/frame/frame_test.go
package frame

import (
	"bytes"
	"encoding/hex"
	"errors"
	"testing"
)

func TestLegacyWriteCmdGoldenVector(t *testing.T) {
	enc := NewEncoder(MsgWriteCmd)
	enc.EncodeVarint(1, 17)
	enc.EncodeVarint(2, 9)
	enc.EncodeBytes(3, []byte{0x01, 0x03, 0x00, 0x00})
	enc.EncodeVarint(4, 8)
	enc.EncodeVarint(5, 33)
	enc.EncodeVarint(6, 2500)

	want, err := hex.DecodeString("06081110091a04010300002008282130c413")
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(enc.Bytes(), want) {
		t.Fatalf("WriteCmd wire = %x, want %x", enc.Bytes(), want)
	}
}

// Test vector from protocol-spec.md §3.2 Hello
func TestHelloWireFormat(t *testing.T) {
	// Encode Hello
	enc := NewEncoder(MsgHello)
	enc.EncodeString(1, "esp32-30eda0a9a808") // device_id
	enc.EncodeString(2, "4.0.0")              // firmware_version
	enc.EncodeString(3, "ESP32S3")            // model
	enc.EncodeVarint(4, 1)                    // channel_count

	wire := enc.Bytes()
	t.Logf("Encoded: %s", hex.EncodeToString(wire))

	// Decode and verify
	dec, err := NewDecoder(wire)
	if err != nil {
		t.Fatalf("decode error: %v", err)
	}
	if dec.MsgType() != MsgHello {
		t.Fatalf("msg type: got 0x%02X, want 0x%02X", dec.MsgType(), MsgHello)
	}

	// Parse all fields
	gotDeviceID := ""
	gotFWVer := ""
	gotModel := ""
	gotChCount := uint64(0)

	for {
		field, err := dec.NextField()
		if err != nil {
			break
		}
		switch field.FieldNum {
		case 1:
			gotDeviceID = GetString(field)
		case 2:
			gotFWVer = GetString(field)
		case 3:
			gotModel = GetString(field)
		case 4:
			gotChCount = GetUint64(field)
		}
	}

	if gotDeviceID != "esp32-30eda0a9a808" {
		t.Errorf("device_id: got %q, want %q", gotDeviceID, "esp32-30eda0a9a808")
	}
	if gotFWVer != "4.0.0" {
		t.Errorf("firmware_version: got %q, want %q", gotFWVer, "4.0.0")
	}
	if gotModel != "ESP32S3" {
		t.Errorf("model: got %q, want %q", gotModel, "ESP32S3")
	}
	if gotChCount != 1 {
		t.Errorf("channel_count: got %d, want 1", gotChCount)
	}

	t.Logf("Hello encode/decode PASS: device=%s fw=%s model=%s ch=%d",
		gotDeviceID, gotFWVer, gotModel, gotChCount)
}

// Test vector from protocol-spec.md §3.4 DataReport
func TestDataReportWireFormat(t *testing.T) {
	// Encode DataReport
	enc := NewEncoder(MsgDataRpt)
	enc.EncodeVarint(1, 5)                                               // channel_id=5
	enc.EncodeVarint(2, 100000000)                                       // timestamp_us=100000000
	enc.EncodeVarint(3, 1)                                               // sequence=1
	enc.EncodeBytes(4, []byte{0x00, 0x41, 0x6e, 0xeb, 0x67, 0x32, 0x00}) // raw_data

	wire := enc.Bytes()
	t.Logf("Encoded: %x", wire)

	// Decode and verify round-trip
	dec, err := NewDecoder(wire)
	if err != nil {
		t.Fatalf("decode error: %v", err)
	}
	if dec.MsgType() != MsgDataRpt {
		t.Fatalf("msg type: got 0x%02X, want 0x%02X", dec.MsgType(), MsgDataRpt)
	}

	gotChID := uint64(0)
	gotTS := uint64(0)
	gotSeq := uint64(0)
	var gotRaw []byte

	for {
		field, err := dec.NextField()
		if err != nil {
			break
		}
		switch field.FieldNum {
		case 1:
			gotChID = GetUint64(field)
		case 2:
			gotTS = GetUint64(field)
		case 3:
			gotSeq = GetUint64(field)
		case 4:
			gotRaw = GetBytes(field)
		}
	}

	if gotChID != 5 {
		t.Errorf("channel_id: got %d, want 5", gotChID)
	}
	if gotTS != 100000000 {
		t.Errorf("timestamp_us: got %d, want 100000000", gotTS)
	}
	if gotSeq != 1 {
		t.Errorf("sequence: got %d, want 1", gotSeq)
	}
	if len(gotRaw) != 7 {
		t.Fatalf("raw_data len: got %d, want 7", len(gotRaw))
	}
	expectedRaw := []byte{0x00, 0x41, 0x6e, 0xeb, 0x67, 0x32, 0x00}
	for i, b := range gotRaw {
		if b != expectedRaw[i] {
			t.Errorf("raw_data[%d]: got 0x%02X, want 0x%02X", i, b, expectedRaw[i])
		}
	}

	t.Logf("DataReport encode/decode PASS: ch=%d ts=%d seq=%d raw=%x",
		gotChID, gotTS, gotSeq, gotRaw)
}

// Test StatusRpt round-trip (protocol-spec §3.3)
func TestStatusRptWireFormat(t *testing.T) {
	// uptime_sec=3600, status="online", channel_count=2
	enc := NewEncoder(MsgStatusRpt)
	enc.EncodeVarint(1, 3600)     // uptime_sec
	enc.EncodeString(2, "online") // status
	enc.EncodeVarint(3, 2)        // channel_count

	wire := enc.Bytes()
	t.Logf("Encoded: %x", wire)

	// Decode round-trip
	dec, err := NewDecoder(wire)
	if err != nil {
		t.Fatalf("decode error: %v", err)
	}
	if dec.MsgType() != MsgStatusRpt {
		t.Fatalf("msg type: got 0x%02X", dec.MsgType())
	}
	t.Logf("StatusRpt wire format PASS")
}

// Test varint boundary values (P1.2 from acceptance-criteria)
func TestVarintBoundaryValues(t *testing.T) {
	tests := []struct {
		value    uint64
		expected int // expected byte count
	}{
		{0, 1},
		{1, 1},
		{127, 1},
		{128, 2},
		{300, 2},
		{16383, 2},
		{16384, 3},
	}

	for _, tt := range tests {
		enc := NewEncoder(MsgHello)
		enc.EncodeVarint(1, tt.value)
		wire := enc.Bytes()
		// wire[0] = msg type, wire[1] = tag byte, wire[2:] = varint bytes
		varintBytes := len(wire) - 2
		if varintBytes != tt.expected {
			t.Errorf("varint(%d): got %d bytes, want %d", tt.value, varintBytes, tt.expected)
		}

		// Round-trip
		dec, _ := NewDecoder(wire)
		field, err := dec.NextField()
		if err != nil {
			t.Fatalf("varint(%d) decode error: %v", tt.value, err)
		}
		got := GetUint64(field)
		if got != tt.value {
			t.Errorf("varint round-trip: got %d, want %d", got, tt.value)
		}
	}
	t.Logf("Varint boundary values PASS (7 cases)")
}

// Test unknown msg type handling (P1.5)
func TestUnknownMsgType(t *testing.T) {
	wire := []byte{0xFF, 0x08, 0x01} // type=0xFF, field 1 varint=1
	dec, err := NewDecoder(wire)
	if err != nil {
		t.Fatalf("should not error on unknown type: %v", err)
	}
	if dec.MsgType() != 0xFF {
		t.Errorf("msg type: got 0x%02X, want 0xFF", dec.MsgType())
	}
	// Should still be able to parse fields
	field, err := dec.NextField()
	if err != nil {
		t.Fatalf("field parse error: %v", err)
	}
	if GetUint64(field) != 1 {
		t.Errorf("field value: got %d, want 1", GetUint64(field))
	}
	t.Logf("Unknown msg type PASS: parsed without crash")
}

// Test unknown field tag handling (P1.6)
func TestUnknownFieldTag(t *testing.T) {
	enc := NewEncoder(MsgHello)
	enc.EncodeVarint(1, 42) // known field
	// Manually inject unknown field 99
	enc.EncodeVarint(99, 123)

	dec, _ := NewDecoder(enc.Bytes())
	fieldCount := 0
	for {
		field, err := dec.NextField()
		if err != nil {
			break
		}
		fieldCount++
		if field.FieldNum == 1 && GetUint64(field) != 42 {
			t.Errorf("field 1: got %d, want 42", GetUint64(field))
		}
	}
	if fieldCount != 2 {
		t.Errorf("field count: got %d, want 2", fieldCount)
	}
	t.Logf("Unknown field tag PASS: skipped without crash")
}

// Test malformed message handling (P1.7)
func TestMalformedMessages(t *testing.T) {
	tests := []struct {
		name string
		data []byte
	}{
		{"empty", []byte{}},
		{"type_only", []byte{0x01}},
		{"truncated_varint", []byte{0x01, 0x08, 0x80, 0x80}}, // varint never terminates
		{"truncated_length", []byte{0x01, 0x0A, 0x10}},       // string len=16 but no data
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			dec, err := NewDecoder(tt.data)
			if err != nil {
				// Empty frame is expected to error
				return
			}
			// Try to parse - should not panic
			for {
				_, err := dec.NextField()
				if err != nil {
					break
				}
			}
		})
	}
	t.Logf("Malformed messages PASS: no panics")
}

// Test SubFrame encoding for ConfigManifest nested structures
func TestSubFrameEncoding(t *testing.T) {
	// Encode a Template sub-structure: id=1, write_data=[0xE0, 0xB6], read_length=25, delay_ms=10
	sub := SubEncoder()
	sub.EncodeVarint(1, 1)                 // id=1
	sub.EncodeBytes(2, []byte{0xE0, 0xB6}) // write_data
	sub.EncodeVarint(3, 25)                // read_length
	sub.EncodeVarint(4, 10)                // delay_ms

	enc := NewEncoder(MsgConfigMfst)
	enc.EncodeString(1, "manifest-001") // manifest_id
	enc.EncodeSubFrame(2, sub.Bytes())  // templates[0]

	wire := enc.Bytes()
	t.Logf("ConfigMfst with sub-frame: %s", hex.EncodeToString(wire))

	// Decode and verify
	dec, err := NewDecoder(wire)
	if err != nil {
		t.Fatalf("decode error: %v", err)
	}
	if dec.MsgType() != MsgConfigMfst {
		t.Fatalf("msg type: got 0x%02X", dec.MsgType())
	}

	gotManifestID := ""
	var gotTemplateData []byte

	for {
		field, err := dec.NextField()
		if err != nil {
			break
		}
		switch field.FieldNum {
		case 1:
			gotManifestID = GetString(field)
		case 2:
			gotTemplateData = GetBytes(field)
		}
	}

	if gotManifestID != "manifest-001" {
		t.Errorf("manifest_id: got %q", gotManifestID)
	}
	if len(gotTemplateData) == 0 {
		t.Error("template sub-frame is empty")
	}

	// Decode sub-frame bytes directly as field sequence
	t.Logf("Template sub-frame decoded, %d bytes", len(gotTemplateData))
	t.Logf("Sub-frame encoding PASS")
}

// Test all message type constants
func TestAllMessageTypes(t *testing.T) {
	expected := map[uint8]string{
		0x01: "hello",
		0x02: "status_report",
		0x03: "data_report",
		0x04: "config_manifest",
		0x05: "config_result",
		0x06: "write_cmd",
		0x07: "write_response",
		0x08: "ping",
		0x09: "pong",
		0x0A: "ota_cmd",
		0x0B: "ota_progress",
		0x0C: "scan_report",
		0x0D: "scan_request",
		0x0E: "query_request",
		0x0F: "query_response",
		0x10: "config_query",
		0x11: "config_report",
		0x12: "hello_ack",
	}

	for msgType, name := range expected {
		got := MsgTypeName(msgType)
		if got != name {
			t.Errorf("MsgTypeName(0x%02X): got %q, want %q", msgType, got, name)
		}
	}
	t.Logf("All 18 message type names PASS")
}

// Test HelloAck (0x12) encode/decode round-trip
func TestHelloAckWireFormat(t *testing.T) {
	serverTime := uint64(1717200000000) // Unix ms
	features := uint64(0)

	enc := NewEncoder(MsgHelloAck)
	enc.EncodeVarint(1, serverTime) // server_time
	enc.EncodeVarint(2, features)   // features

	wire := enc.Bytes()
	t.Logf("HelloAck encoded: %s", hex.EncodeToString(wire))

	dec, err := NewDecoder(wire)
	if err != nil {
		t.Fatalf("decode error: %v", err)
	}
	if dec.MsgType() != MsgHelloAck {
		t.Fatalf("msg type: got 0x%02X, want 0x%02X", dec.MsgType(), MsgHelloAck)
	}

	var gotServerTime, gotFeatures uint64
	for {
		field, err := dec.NextField()
		if err != nil {
			break
		}
		switch field.FieldNum {
		case 1:
			gotServerTime = GetUint64(field)
		case 2:
			gotFeatures = GetUint64(field)
		}
	}

	if gotServerTime != serverTime {
		t.Errorf("server_time: got %d, want %d", gotServerTime, serverTime)
	}
	if gotFeatures != features {
		t.Errorf("features: got %d, want %d", gotFeatures, features)
	}
	t.Logf("HelloAck encode/decode PASS: server_time=%d features=%d", gotServerTime, gotFeatures)
}

func TestHelloV26HandshakeNonceRoundTrip(t *testing.T) {
	const nonce uint32 = 0xA1B2C3D4
	enc := NewEncoder(MsgHello)
	enc.EncodeVarint(HelloFieldHandshakeNonce, uint64(nonce))

	wire := enc.Bytes()
	if len(wire) < 2 || wire[1] != byte(HelloFieldHandshakeNonce<<3|WireVarint) {
		t.Fatalf("Hello nonce tag: got %x", wire)
	}
	dec, err := NewDecoder(wire)
	if err != nil {
		t.Fatalf("NewDecoder: %v", err)
	}
	field, err := dec.NextField()
	if err != nil {
		t.Fatalf("NextField: %v", err)
	}
	if field.FieldNum != HelloFieldHandshakeNonce || field.WireType != WireVarint || GetUint64(field) != uint64(nonce) {
		t.Fatalf("nonce field: got %#v, want %d", field, nonce)
	}
}

func TestExtendedFieldTagsRoundTrip(t *testing.T) {
	enc := NewEncoder(MsgStatusRpt)
	enc.EncodeVarint(18, 42)
	enc.EncodeBytes(23, []byte("x"))
	wire := enc.Bytes()
	if len(wire) < 8 || wire[1] != 0x90 || wire[2] != 0x01 {
		t.Fatalf("field 18 tag is not canonical: %x", wire)
	}
	dec, err := NewDecoder(wire)
	if err != nil {
		t.Fatalf("NewDecoder: %v", err)
	}
	field, err := dec.NextField()
	if err != nil || field.FieldNum != 18 || GetUint64(field) != 42 {
		t.Fatalf("field 18 round trip: field=%#v err=%v", field, err)
	}
	field, err = dec.NextField()
	if err != nil || field.FieldNum != 23 || string(GetBytes(field)) != "x" {
		t.Fatalf("field 23 round trip: field=%#v err=%v", field, err)
	}
}

func TestHelloAckV26HandshakeNonceRoundTrip(t *testing.T) {
	const nonce uint32 = 0xFFFFFFFF
	enc := NewEncoder(MsgHelloAck)
	enc.EncodeVarint(1, 1)
	enc.EncodeVarint(2, 0)
	enc.EncodeVarint(HelloAckFieldHandshakeNonce, uint64(nonce))

	dec, err := NewDecoder(enc.Bytes())
	if err != nil {
		t.Fatalf("NewDecoder: %v", err)
	}
	var got uint64
	for {
		field, err := dec.NextField()
		if errors.Is(err, ErrEndOfFrame) {
			break
		}
		if err != nil {
			t.Fatalf("NextField: %v", err)
		}
		if field.FieldNum == HelloAckFieldHandshakeNonce {
			got = GetUint64(field)
		}
	}
	if got != uint64(nonce) {
		t.Fatalf("echoed nonce: got %d, want %d", got, nonce)
	}
}

func TestDecoderRejectsOverflowingVarint(t *testing.T) {
	wire := []byte{MsgHello, byte(HelloFieldHandshakeNonce << 3)}
	for range 9 {
		wire = append(wire, 0x80)
	}
	wire = append(wire, 0x02) // tenth byte may carry only bit zero for uint64

	dec, err := NewDecoder(wire)
	if err != nil {
		t.Fatalf("NewDecoder: %v", err)
	}
	if _, err := dec.NextField(); err == nil {
		t.Fatal("overflowing uint64 varint was accepted")
	}
}

func TestDecoderRejectsNonCanonicalVarint(t *testing.T) {
	wire := []byte{
		MsgHello, byte(HelloFieldHandshakeNonce << 3), 0x80, 0x00,
	}
	dec, err := NewDecoder(wire)
	if err != nil {
		t.Fatalf("NewDecoder: %v", err)
	}
	if _, err := dec.NextField(); err == nil {
		t.Fatal("non-canonical zero varint was accepted")
	}
}

// === Benchmarks (P2.x from acceptance-criteria) ===

// BenchmarkDecodeHello measures decode performance for typical Hello message
func BenchmarkDecodeHello(b *testing.B) {
	// Pre-encode a typical Hello message
	enc := NewEncoder(MsgHello)
	enc.EncodeString(1, "esp32c6_404CCA57B7BC")
	enc.EncodeString(2, "2.0.0")
	enc.EncodeString(3, "ESP32C6")
	enc.EncodeVarint(4, 4)
	wire := enc.Bytes()

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		dec, _ := NewDecoder(wire)
		for {
			_, err := dec.NextField()
			if err != nil {
				break
			}
		}
	}
}

// BenchmarkEncodeHello measures encode performance
func BenchmarkEncodeHello(b *testing.B) {
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		enc := NewEncoder(MsgHello)
		enc.EncodeString(1, "esp32c6_404CCA57B7BC")
		enc.EncodeString(2, "2.0.0")
		enc.EncodeString(3, "ESP32C6")
		enc.EncodeVarint(4, 4)
		_ = enc.Bytes()
	}
}

// BenchmarkDecodeDataReport measures decode performance for typical DataReport
func BenchmarkDecodeDataReport(b *testing.B) {
	// Pre-encode a typical DataReport (7 bytes raw data)
	enc := NewEncoder(MsgDataRpt)
	enc.EncodeVarint(1, 5)
	enc.EncodeVarint(2, 1717200000000)
	enc.EncodeVarint(3, 12345)
	enc.EncodeBytes(4, []byte{0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07})
	wire := enc.Bytes()

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		dec, _ := NewDecoder(wire)
		for {
			_, err := dec.NextField()
			if err != nil {
				break
			}
		}
	}
}

// BenchmarkEncodeDataReport measures encode performance
func BenchmarkEncodeDataReport(b *testing.B) {
	rawData := []byte{0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		enc := NewEncoder(MsgDataRpt)
		enc.EncodeVarint(1, 5)
		enc.EncodeVarint(2, 1717200000000)
		enc.EncodeVarint(3, 12345)
		enc.EncodeBytes(4, rawData)
		_ = enc.Bytes()
	}
}

// BenchmarkDecodeConfigManifest measures decode for complex nested message
func BenchmarkDecodeConfigManifest(b *testing.B) {
	// Pre-encode a ConfigManifest with 4 templates
	enc := NewEncoder(MsgConfigMfst)
	enc.EncodeString(1, "cfg-001")
	for i := 0; i < 4; i++ {
		sub := SubEncoder()
		sub.EncodeVarint(1, uint64(i+1))
		sub.EncodeBytes(2, []byte{0xE0, 0xB6})
		sub.EncodeVarint(3, 25)
		sub.EncodeVarint(4, 10)
		enc.EncodeSubFrame(2, sub.Bytes())
	}
	wire := enc.Bytes()

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		dec, _ := NewDecoder(wire)
		for {
			_, err := dec.NextField()
			if err != nil {
				break
			}
		}
	}
}

// Fuzz test for frame decoder (N3.3)
func FuzzDecoder(f *testing.F) {
	// Seed corpus
	f.Add([]byte{0x01, 0x0A, 0x05, 'h', 'e', 'l', 'l', 'o'}) // Hello with short string
	f.Add([]byte{0x03, 0x08, 0x01})                          // DataReport with varint 1
	f.Add([]byte{0xFF, 0x08, 0x7F})                          // Unknown type, varint 127
	f.Add([]byte{})                                          // empty
	f.Add([]byte{0x01, 0x80, 0x80, 0x80, 0x80})              // malformed varint

	f.Fuzz(func(t *testing.T, data []byte) {
		// Should never panic
		dec, err := NewDecoder(data)
		if err != nil {
			return
		}
		// Consume all fields safely
		for {
			field, err := dec.NextField()
			if err != nil {
				break
			}
			// Try to read all value types - should not panic
			_ = GetUint64(field)
			_ = GetString(field)
			_ = GetBool(field)
		}
	})
}

// Fuzz test for encoder round-trip
func FuzzEncoderRoundTrip(f *testing.F) {
	f.Add(uint8(0x01), uint64(42), "hello", true)
	f.Add(uint8(0x03), uint64(1), "test", false)
	f.Add(uint8(0x0B), uint64(100), "ota-task", true)

	f.Fuzz(func(t *testing.T, msgType uint8, value uint64, str string, flag bool) {
		// Encode
		enc := NewEncoder(msgType)
		enc.EncodeVarint(1, value)
		enc.EncodeString(2, str)
		enc.EncodeBool(3, flag)
		wire := enc.Bytes()

		// Decode and verify
		dec, err := NewDecoder(wire)
		if err != nil {
			t.Fatalf("decode error: %v", err)
		}
		if dec.MsgType() != msgType {
			t.Errorf("msg type: got 0x%02X, want 0x%02X", dec.MsgType(), msgType)
		}

		// Parse all fields
		var gotVal uint64
		var gotStr string
		var gotFlag bool
		for {
			field, err := dec.NextField()
			if err != nil {
				break
			}
			switch field.FieldNum {
			case 1:
				gotVal = GetUint64(field)
			case 2:
				gotStr = GetString(field)
			case 3:
				gotFlag = GetBool(field)
			}
		}

		if gotVal != value {
			t.Errorf("value: got %d, want %d", gotVal, value)
		}
		if gotStr != str {
			t.Errorf("str: got %q, want %q", gotStr, str)
		}
		if gotFlag != flag {
			t.Errorf("flag: got %v, want %v", gotFlag, flag)
		}
	})
}
